File: src/python/parallelism.py
import requests
import json 
from multiprocessing import Process
import random
import time
from tqdm import tqdm
from multiprocessing import Pool

def http_request(request_count):
    # for i in tqdm(range(request_count)):
        id_num = random.randint(1,100)
        res = requests.get('https://dummyjson.com/todos/' + str(id_num))
        response = json.loads(res.text)
        # print(response)

            
if __name__ == '__main__':
    start_time = time.perf_counter()
    with Pool() as pool:
        result = pool.map(http_request, range(0,1000))
    finish_time = time.perf_counter()
    print("Program finished in {} seconds - using multiprocessing".format(finish_time-start_time))
    print("---")

    start_time = time.perf_counter()
    for i in range(1000):
        http_request(1000)
    finish_time = time.perf_counter()
    print("Program finished in {} seconds".format(finish_time-start_time))
    print("---")

File: src/go/parallelism.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"sync"
	"time"
)

func httpRequest(requestCount int) {
	idNum := rand.Intn(100) + 1

	resp, err := http.Get(fmt.Sprintf("https://dummyjson.com/todos/%d", idNum))
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	var result map[string]interface{}
	err = json.Unmarshal(body, &result)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Uncomment the following line to print the response
	// fmt.Println(result)
}

func main() {
	numRequests := 1000

	// Parallel execution using goroutines and a WaitGroup
	startTime := time.Now()
	var wg sync.WaitGroup

	for i := 0; i < numRequests; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			httpRequest(1000)
		}()
	}

	wg.Wait()
	elapsedTime := time.Since(startTime)
	fmt.Printf("Program finished in %s - using goroutines\n", elapsedTime)
	fmt.Println("---")

	// Sequential execution
	startTime = time.Now()
	for i := 0; i < numRequests; i++ {
		httpRequest(1000)
	}
	elapsedTime = time.Since(startTime)
	fmt.Printf("Program finished in %s\n", elapsedTime)
	fmt.Println("---")
}
